pandas/core/missing.py

```diff
@@ -973,3 +981,86 @@ def _rolling_window(a: npt.NDArray[np.bool_], window: int) -> npt.NDArray[np.boo
     shape = a.shape[:-1] + (a.shape[-1] - window + 1, window)
     strides = a.strides + (a.strides[-1],)
     return np.lib.stride_tricks.as_strided(a, shape=shape, strides=strides)
+
+
+def _can_hold_element(values, element: Any) -> bool:
+    """
+    Expanded version of core.dtypes.cast.can_hold_element
+    """
+    element = extract_array(element, extract_numpy=True)
+    return can_hold_element(values, element)
+
+
+def fillna_ea_array(values, value, limit=None, inplace: bool = False, downcast=None):
+    """
+    Fillna logic for ExtensionArrays.
+
+    Dispatches to the EA.fillna method (in which case downcast is currently
+    ignored), except for datetime64 in which case fallback to object dtype
+    is currently allowed.
+    """
+    if not _can_hold_element(values, value) and values.dtype.kind == "M":
```

Comment:
Then we don't have tests for it in any case, since the tests are passing.

Now, checking for period, it seems to raise for both of array and Series:

```
In [5]: arr = pd.period_range("2012", periods=3).array

In [6]: arr[0] = None

In [8]: arr.fillna("?")
...
TypeError: value should be a 'Period', 'NaT', or array of those. Got 'str' instead.

In [9]: pd.Series(arr).fillna("?")
...
TypeError: value should be a 'Period', 'NaT', or array of those. Got 'str' instead.

```

while indeed for DatetimeArray, we coerce to object for Series:
```
In [10]: arr = pd.date_range("2012", periods=3).array

In [11]: arr[0] = None

In [12]: arr.fillna("?")
..
TypeError: value should be a 'Timestamp', 'NaT', or array of those. Got 'str' instead.

In [13]: pd.Series(arr).fillna("?")
Out[13]: 
0                      ?
1    2012-01-02 00:00:00
2    2012-01-03 00:00:00
dtype: object

```